internal/bgp/bgp.go

```diff
@@ -155,8 +161,21 @@ func (s *Session) connect() error {
 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
 	defer cancel()
 
-	var d net.Dialer
-	conn, err := d.DialContext(ctx, "tcp", s.addr)
+	laddr, err := net.ResolveTCPAddr("tcp", net.JoinHostPort("0.0.0.0", "0"))
+	var conn net.Conn
+	d := TCPDialer{
+		Dialer: net.Dialer{
+			LocalAddr: laddr,
+			Timeout:   10 * time.Second,
+		},
+		AuthPassword: s.password,
+	}
+
+	d.TTL = 10
+	tcphost, portstr, err := net.SplitHostPort(s.addr)
+	port, err := strconv.Atoi(portstr)
+	conn, err = d.DialTCP(tcphost, port)
```

Comment:
The code you replaced used `ctx` to pass in a timeout to the overall connection logic. The replacement code no longer does that, so we've lost functionality here. Can you plumb `ctx` into DialTCP, and use the deadline it provides for various timeouts?